java/src/main/native/src/TableJni.cpp

```diff
@@ -1455,109 +1568,143 @@ JNIEXPORT void JNICALL Java_ai_rapids_cudf_Table_readArrowIPCEnd(JNIEnv *env, jc
-                                                                  cudf::null_equality::UNEQUAL);
-
-    return cudf::jni::convert_table_for_return(env, result);
+    auto left_in_table = reinterpret_cast<cudf::table_view *>(j_left_table);
+    auto right_in_table = reinterpret_cast<cudf::table_view *>(j_right_table);
+    cudf::jni::native_jintArray left_key_indices(env, j_left_key_indices);
+    auto left_keys_table = cudf::jni::get_keys_table(left_in_table, left_key_indices);
+    left_key_indices.cancel();
+    cudf::jni::native_jintArray right_key_indices(env, j_right_key_indices);
+    auto right_keys_table = cudf::jni::get_keys_table(right_in_table, right_key_indices);
+    auto nulleq = compare_nulls_equal ? cudf::null_equality::EQUAL : cudf::null_equality::UNEQUAL;
+
+    // compute gather maps for the left and right tables that can produce the join result rows
+    auto join_maps = cudf::inner_join(left_keys_table, right_keys_table, nulleq);
+    CUDF_EXPECTS(join_maps.first->size() <= std::numeric_limits<cudf::size_type>::max(),
+                 "join result exceeds maximum column length");
+    auto num_join_rows = static_cast<cudf::size_type>(join_maps.first->size());
+
+    // compute the join result rows for the left table columns
+    auto left_gather_col = cudf::column_view(cudf::data_type{cudf::type_id::INT32}, num_join_rows,
+                                             join_maps.first->data());
+    auto left_out_table =
+        cudf::gather(*left_in_table, left_gather_col, cudf::out_of_bounds_policy::DONT_CHECK);
+
+    // compute the join result rows for the right table columns
+    auto right_non_keys_table = cudf::jni::get_non_keys_table(right_in_table, right_key_indices);
+    right_key_indices.cancel();
+    auto right_gather_col = cudf::column_view(cudf::data_type{cudf::type_id::INT32}, num_join_rows,
+                                              join_maps.second->data());
+    auto right_out_table = cudf::gather(right_non_keys_table, right_gather_col,
+                                        cudf::out_of_bounds_policy::DONT_CHECK);
+
+    return cudf::jni::combine_join_results(env, *left_out_table, *right_out_table);
   }
   CATCH_STD(env, NULL);
 }
 
 JNIEXPORT jlongArray JNICALL Java_ai_rapids_cudf_Table_fullJoin(
-    JNIEnv *env, jclass, jlong left_table, jintArray left_col_join_indices, jlong right_table,
-    jintArray right_col_join_indices, jboolean compare_nulls_equal) {
-  JNI_NULL_CHECK(env, left_table, "left_table is null", NULL);
-  JNI_NULL_CHECK(env, left_col_join_indices, "left_col_join_indices is null", NULL);
-  JNI_NULL_CHECK(env, right_table, "right_table is null", NULL);
-  JNI_NULL_CHECK(env, right_col_join_indices, "right_col_join_indices is null", NULL);
+    JNIEnv *env, jclass, jlong j_left_table, jintArray j_left_key_indices, jlong j_right_table,
+    jintArray j_right_key_indices, jboolean compare_nulls_equal) {
+  JNI_NULL_CHECK(env, j_left_table, "left_table is null", NULL);
```

Comment:
If we are planning on keeping these APIs, should we make a single template function like we did for the new APIs?

Reply:
I originally went down that path, but then full join ended up being significantly different which spoiled a lot of the reuse potential and ended up making it complicated to reuse in a template without adding unnecessary overhead.  I can still do templates to reuse code between left and inner joins since those are very close, just changing the join function and `out_of_bounds_policy` settings.  Could do something similar with semi and anti join, although those were duplicated before this PR and I didn't significantly alter them.